SQL_DDL_HW

h6bi opened the pull request.

h6bi committed:
uploaded WEEK 4 HW

KseniyaLazarchik reviewed (commented):
Hi @h6bi 
Good job
Some issues:

1. You can add previously DROP DATABASE IF EXISTS an than CREATE DATABASE 
For making code rerunable
Use CREATE SCHEMA IF NOT EXISTS 

2. Second run will failure for 
ALTER TABLE training.additional_services
    ADD CONSTRAINT price_usd_check CHECK (price_USD > 0);

ALTER TABLE training.companies
    ADD CONSTRAINT possible_legal_forms CHECK (legal_form IN ('SP', 'GP', 'LP', 'LLC', 'Ltd', 'Inc', 'NPO', 'PLC', 'BO', 'JV'));
You can try 
DO $$
BEGIN
    IF NOT EXISTS (
        SELECT *
        FROM information_schema.check_constraints cc
        JOIN information_schema.table_constraints tc
            ON cc.constraint_name = tc.constraint_name
        WHERE tc.table_schema = 'training'
          AND tc.table_name = 'additional_services'
          AND cc.constraint_name = 'price_usd_check'
    ) THEN
        ALTER TABLE training.additional_services
            ADD CONSTRAINT price_usd_check CHECK (price_USD > 0);
    END IF;
END
$$;
...

3. INSERTs is not rerunable (use or WHERE NOT EXISTS, or ON CONFLICT DO NOTHING (UPDATE)) 

4. Don't use hardcoded job_ID in training.interviews

5. Don't see last part of the task: about adding record_ts column to all tables

6. GENERATED ALWAYS is missing

h6bi committed:
Correction #1

h6bi commented:
Hi @KseniyaLazarchik :) thank you for the feedback!

Commited the following corrections:
- added DROP DATABASE IF EXISTS and CREATE SCHEMA IF NOT EXISTS
- wrapped ALTER TABLE statements in PL/pgsql blocks for rerunability
- added WHERE NOT EXISTS to INSERT statements for rerunability
- fixed hardcoded ids
- added record_ts column to all tables

I have a question about GENERATED ALWAYS AS part: I don't see a reason and opportunity to use it for any column in my current DB schema. Can you please advise, where I should use it, and if I have to modify my schema accordingly?

Thank you in advance!

KseniyaLazarchik commented:
Hi @h6bi 
Good 
You can merge

KseniyaLazarchik commented:
> I have a question about GENERATED ALWAYS AS part: I don't see a reason and opportunity to use it for any column in my current DB schema. Can you please advise, where I should use it, and if I have to modify my schema accordingly?

As for this point - this is one of the requirements of the task, if you think that there is nowhere to apply it - it's your right, I will accept the task as is
